interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,384 @@
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == metadata.Snap && prompt.Interface == metadata.Interface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			// Although the prompt itself has not changed, re-record a notice
+			// to re-notify clients to respond to this request. A client may
+			// have replied with a malformed response and not retried after
+			// receiving the error, so this notice encourages it to try again
+			// if the user retries the operation.
+			pdb.notifyPrompt(metadata.User, prompt.ID, nil)
+			return prompt, true
+		}
+	}
+
+	id := pdb.nextID()
+	timestamp := time.Now()
+	prompt := &Prompt{
+		ID:           id,
+		Timestamp:    timestamp,
+		Snap:         metadata.Snap,
+		Interface:    metadata.Interface,
+		Constraints:  constraints,
+		listenerReqs: []*listener.Request{listenerReq},
+	}
+	userEntry.ByID[id] = prompt
+	pdb.notifyPrompt(metadata.User, id, nil)
+	return prompt, false
+}
+
+// Prompts returns a slice of all outstanding prompts.
+func (pdb *PromptDB) Prompts(user uint32) []*Prompt {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return []*Prompt{}
+	}
+	prompts := make([]*Prompt, 0, len(userEntry.ByID))
+	for _, prompt := range userEntry.ByID {
+		prompts = append(prompts, prompt)
+	}
+	return prompts
+}
+
+// PromptWithID returns the prompt with the given ID for the given user.
+func (pdb *PromptDB) PromptWithID(user uint32, id string) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	_, prompt, err := pdb.promptWithID(user, id)
+	return prompt, err
+}
+
+// promptWithID returns the user entry for the given user and the prompt with
+// the given ID for the that user.
+//
+// The caller should hold the prompt DB lock.
+func (pdb *PromptDB) promptWithID(user uint32, id string) (*userPromptDB, *Prompt, error) {
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil, nil, ErrNotFound
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, nil, ErrNotFound
+	}
+	return userEntry, prompt, nil
+}
+
+// Reply resolves the prompt with the given ID using the given outcome by
+// sending a reply to all associated listener requests, then removing the
+// prompt from the prompt DB.
+//
+// Records a notice for the prompt, and returns the prompt's former contents.
+func (pdb *PromptDB) Reply(user uint32, id string, outcome prompting.OutcomeType) (*Prompt, error) {
+	allow, err := outcome.AsBool()
+	if err != nil {
+		return nil, err
+	}
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, prompt, err := pdb.promptWithID(user, id)
+	if err != nil {
+		return nil, err
+	}
+	for _, listenerReq := range prompt.listenerReqs {
+		if err := sendReply(listenerReq, allow); err != nil {
+			return nil, err
+		}
+	}
+	delete(userEntry.ByID, id)
```

Comment:
Should we ever reallocate and garbage collect the map so that it does not hold a lot of memory from the past?

Reply:
Can try to make this better. If we store sub-maps by snap and interface, we could provide more opportunities for setting maps to `nil` (allowing GC). But if loads of prompts are coming from the snap using the same interface, it doesn't help much.

That suggestion would help with speed of looking for matching prompts, though.